adiciona tratativa para grading students

Morkhusz opened the pull request:
Proposta para a issue #15

Morkhusz committed:
adiciona tratativa para grading students

nawarian commented on src/GradingStudents/GradingStudents.php:
```diff
@@ -0,0 +1,34 @@
+
+class GradingStudents
+{
+    public function roundGrades($grades)
```
> grades: an array of integers representing grades before rounding

O input deveria ser um array, não string. Aproveita, e adiciona um type hint aí ;)

Morkhusz replied:
Acho que a parte da descrição da função acabou dando uma confundida nessa parte, no programa do hacker rank realmente é um array mas no *input format* mostra o formato certo que seria:
>The first line contains a single integer, **n**, the number of students.
Each line **i** of the **n** subsequent lines contains a single integer, **grades[i]**, denoting student 's grade.

Por isso o input está como string =s

nawarian commented on src/GradingStudents/GradingStudents.php:
```diff
@@ -0,0 +1,34 @@
+    {
+        $grades = explode(PHP_EOL, $grades);
+        $students = (int) array_shift($grades);
+        $this->validateInput($students, $grades);
```
Como a validação é curta e rápida, pode ser trazida pra mesma função. Desta maneira você evita fazer cópia de valores de memória.

nawarian commented on src/GradingStudents/GradingStudents.php:
```diff
@@ -0,0 +1,34 @@
+        $students = (int) array_shift($grades);
+        $this->validateInput($students, $grades);
+
+        return implode(PHP_EOL, array_map(function ($grade) {
```
>The function should return an integer array consisting of rounded grades.

O retorno também deveria ser um array, portanto o `implode()` aqui não é necessário. Você também pode adicionar type hints na assinatura da função.

Morkhusz replied:
Novamente isso ficou confuso por conta da descrição da função =/
O dado para output é o seguinte:
>**Output Format**
For each grades[i], print the rounded grade on a new line.

nawarian commented on src/GradingStudents/GradingStudents.php:
```diff
@@ -0,0 +1,34 @@
+        $this->validateInput($students, $grades);
+
+        return implode(PHP_EOL, array_map(function ($grade) {
+            if ($grade < 0 || $grade > 100) {
```
Você pode evitar executar este IF várias vezes ao validar apenas uma vez antes do seu loop (array_map).

Você pode utilizar as funções `min()` e `max()` para obter o menor e maior valores neste array. Se `min($grades)` for menor que 0 ou `max($grades)` for maior que 100, você lança a exceção.

Refs:
- https://www.php.net/manual/en/function.min.php
- https://www.php.net/manual/en/function.max

nawarian commented on src/GradingStudents/GradingStudents.php:
```diff
@@ -0,0 +1,34 @@
+        $students = (int) array_shift($grades);
+        $this->validateInput($students, $grades);
+
+        return implode(PHP_EOL, array_map(function ($grade) {
```
Apesar de parecer elegante, este `array_map()` poderia ser um simples foreach. Você evitaria `N` fcalls, passar o array por cópia e o type check de `$grade` a cada iteração.

Morkhusz committed:
Implementa melhorias sugeridas na avaicacao do PR

Remove metodo de validacao do input e traz para a proria funcao principal
Remove utilizacao dos explode e implodes tratando a string de uma so vez

Morkhusz commented on src/GradingStudents/GradingStudents.php:
```diff
@@ -0,0 +1,36 @@
+
+namespace App\GradingStudents;
+
+use InvalidArgumentException;
```
```suggestion
use InvalidArgumentException123;
```
